qiskit/providers/ibmq/api_v2/clients/websocket.py

```diff
@@ -150,64 +153,81 @@ def get_job_status(self, job_id, timeout=None):
+            original_timeout = timeout
+            start_time = time.time()
+            last_status = None
+
+            try:
+                # Read messages from the server until the connection is closed or
+                # a timeout has been reached.
+                while True:
+                    try:
+                        with warnings.catch_warnings():
+                            # Suppress websockets deprecation warnings until the fix is available
+                            warnings.filterwarnings("ignore", category=DeprecationWarning)
+                            if timeout:
+                                response_raw = yield from asyncio.wait_for(
+                                    websocket.recv(), timeout=timeout)
+
+                                # Decrease the timeout, with a 5-second grace period.
+                                elapsed_time = time.time() - start_time
+                                timeout = max(5, int(original_timeout - elapsed_time))
+                            else:
+                                response_raw = yield from websocket.recv()
+                        logger.debug('Received message from websocket: %s',
+                                     response_raw)
+
+                        response = WebsocketMessage.from_bytes(response_raw)
+                        last_status = response.data
+
+                        job_status = response.data.get('status')
+                        if (job_status and
+                                ApiJobStatus(job_status) in API_JOB_FINAL_STATES):
+                            break
+
+                    except futures.TimeoutError:
+                        # Timeout during our wait.
+                        raise WebsocketTimeoutError('Timeout reached') from None
+                    except ConnectionClosed as ex:
+                        # From the API:
+                        # 4001: closed due to an internal errors
+                        # 4002: closed on purpose (no more updates to send)
+                        # 4003: closed due to job not found.
+                        connection_closed_ex_code = ex.code
+                        message = 'Unexpected error'
+
+                        if ex.code == 4001:
+                            message = 'Internal server error'
+                        elif ex.code == 4002:
+                            break
+                        elif ex.code == 4003:
+                            message = 'Job id not found'
+                        raise WebsocketError('Connection with websocket closed '
+                                             'unexpectedly: {}'.format(message)) from ex
+            except WebsocketError as ex:
+                # Maxed out the number of retries.
+                if current_retry >= retry_max:
+                    raise ex
+                # Job id not found. No point of retrying.
+                if connection_closed_ex_code == 4003:
+                    raise ex
+
```

Comment:
Most logical way I could think of retrying was to catch the exception and decide whether a retry attempt should be made.

Reply:
You should be able to just check for the error code inside the `except` block of 197 and do a reconnect instead of raising. If you still end up doing this additional `try`, you might want to break the block into helper functions to make the code more readable.

Reply:
I was thinking of reconnecting within the `except` block of 197. But if the connection is reestablished inside the `except` block, wouldn't the `finally` block close the connection right after, since it always gets executed after?

Reply:
Yes, that makes perfect sense! I was thinking of things from a different perspective, but I think the way you mentioned might keep things more simplified and tidy.